crates/proc_macro_api/src/lib.rs

```diff
@@ -108,4 +121,61 @@ impl ProcMacroClient {
             }
         }
     }
+
+    // This is used inside self.read_version() to locate the ".rustc" section
+    // from a proc macro crate's binary file.
+    fn read_section(&self, dylib_binary: &[u8], section_name: &str) -> &[u8] {
+        BinaryFile::parse(dylib_binary)
+            .unwrap()
+            .section_by_name(section_name)
+            .unwrap()
+            .data()
+            .unwrap()
+    }
+
+    // Check the version of rustc that was used to compile a proc macro crate's
+    // binary file.
+    // A proc macro crate binary's ".rustc" section has following byte layout:
+    // * [b'r',b'u',b's',b't',0,0,0,5] is the first 8 bytes
+    // * ff060000 734e6150 is followed, it's the snappy format magic bytes,
+    //   means bytes from here(including this sequence) are compressed in
+    //   snappy compression format. Version info is here inside, so decompress
+    //   this.
+    // The bytes you get after decompressing the snappy format portion has
+    // following layout:
+    // * [b'r',b'u',b's',b't',0,0,0,5] is the first 8 bytes(again)
+    // * [crate root bytes] next 4 bytes is to store crate root position,
+    //   according to rustc's source code comment
+    // * [length byte] next 1 byte tells us how many bytes we should read next
+    //   for the version string's utf8 bytes
+    // * [version string bytes encoded in utf8] <- GET THIS BOI
+    // * [some more bytes that we don really care but still there] :-)
+    // Check this issue for more about the bytes layout:
+    // https://github.com/rust-analyzer/rust-analyzer/issues/6174
+    fn read_version(&self, dylib_path: &Path) -> String {
+        let dylib_binary = fsread(dylib_path).unwrap();
+
+        let dot_rustc = self.read_section(&dylib_binary, ".rustc");
+
+        let snappy_portion = &dot_rustc[8..];
+
+        let mut snappy_decoder = SnapDecoder::new(snappy_portion);
+
+        // the bytes before version string bytes, so this basically is:
+        // 8 bytes for [b'r',b'u',b's',b't',0,0,0,5]
+        // 4 bytes for [crate root bytes]
+        // 1 byte for length of version string
+        // so 13 bytes in total, and we should check the 13th byte
+        // to know the length
+        let mut bytes_before_version = [0u8; 13];
+        snappy_decoder
+            .read_exact(&mut bytes_before_version)
+            .unwrap();
+        let length = bytes_before_version[12]; // what? can't use -1 indexing?
```

Comment:
This should probably also check that the metadata version is 5.

Reply:
Or rather, it should be checked before decompression.

Reply:
You mean, the last byte `5` in 8 bytes sequence `r u s t 0 0 0 5`?

Reply:
Yes

Reply:
I think this is resolved by the condition check `if header != EXPECTED_HEADER {...}` that I added in later commit? @bjorn3

Reply:
> I think this is resolved by the condition check if header != EXPECTED_HEADER {...} that I added in later commit?

Yes, I think so .